File: samples/tutorials-native/tutorials-3d/tutorial-4/javaBuilder/src/main/java/com/scichart/tutorial/MainActivity.java
package com.scichart.tutorial;

import android.os.Bundle;
import android.util.Log;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;

import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting3d.model.dataSeries.xyz.XyzDataSeries3D;
import com.scichart.charting3d.modifiers.OrbitModifier3D;
import com.scichart.charting3d.modifiers.PinchZoomModifier3D;
import com.scichart.charting3d.modifiers.TooltipModifier3D;
import com.scichart.charting3d.modifiers.ZoomExtentsModifier3D;
import com.scichart.charting3d.visuals.SciChartSurface3D;
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D;
import com.scichart.charting3d.visuals.renderableSeries.scatter.ScatterRenderableSeries3D;
import com.scichart.core.framework.UpdateSuspender;
import com.scichart.core.model.DoubleValues;
import com.scichart.extensions3d.builders.SciChart3DBuilder;

import java.util.Collections;
import java.util.Random;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    // <DataSeriesSetup>
    private int pointsCount = 200;

    private ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
    private ScheduledFuture schedule;

    private SciChartSurface3D surface;

    private final DoubleValues xValues = new DoubleValues();
    private final DoubleValues yValues = new DoubleValues();
    private final DoubleValues zValues = new DoubleValues();

    private final XyzDataSeries3D dataSeries = new XyzDataSeries3D(Double.class, Double.class, Double.class);
    // </DataSeriesSetup>

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        try {
            SciChartSurface.setRuntimeLicenseKey("");
        } catch (Exception e) {
            Log.e("SciChart", "Error during setting license key", e);
        }

        LinearLayout chartLayout = findViewById(R.id.chart_layout);

        surface = new SciChartSurface3D(this);
        chartLayout.addView(surface);

        SciChart3DBuilder.init(this);
        final SciChart3DBuilder sciChartBuilder = SciChart3DBuilder.instance();

        // <AppendDataSeries>
        for (int i = 0; i < pointsCount; i++) {
            xValues.add(getGaussianRandomNumber(5.0, 1.5));
            yValues.add(getGaussianRandomNumber(5.0, 1.5));
            zValues.add(getGaussianRandomNumber(5.0, 1.5));
        }
        dataSeries.append(xValues, yValues, zValues);
        // </AppendDataSeries>

        final SpherePointMarker3D pointMarker = sciChartBuilder.newSpherePointMarker3D()
                .withFill(0xFF32CD32)
                .withSize(10.0f)
                .build();

        final ScatterRenderableSeries3D rSeries = sciChartBuilder.newScatterSeries3D()
                .withDataSeries(dataSeries)
                .withPointMarker(pointMarker)
                .build();

        final TooltipModifier3D tooltipModifier = new TooltipModifier3D();
        tooltipModifier.setExecuteOnPointerCount(2);

        UpdateSuspender.using(surface, () -> {
            surface.setXAxis(sciChartBuilder.newNumericAxis3D().build());
            surface.setYAxis(sciChartBuilder.newNumericAxis3D().build());
            surface.setZAxis(sciChartBuilder.newNumericAxis3D().build());
            Collections.addAll(surface.getRenderableSeries(), rSeries);
            Collections.addAll(surface.getChartModifiers(),
                    new OrbitModifier3D(),
                    new ZoomExtentsModifier3D(),
                    new PinchZoomModifier3D(),
                    tooltipModifier
            );
        });

        // <CreateTimer>
        schedule = scheduledExecutorService.scheduleWithFixedDelay(updateData, 0, 10, TimeUnit.MILLISECONDS);
        // </CreateTimer>
    }

    // <UpdateData>
    private Random random = new Random();

    private final Runnable updateData = () -> {
        for (int i = 0; i < pointsCount; i++) {
            double xValue = xValues.get(i) + random.nextDouble() - 0.5;
            double yValue = yValues.get(i) + random.nextDouble() - 0.5;
            double zValue = zValues.get(i) + random.nextDouble() - 0.5;

            xValues.set(i, xValue);
            yValues.set(i, yValue);
            zValues.set(i, zValue);
        }

        UpdateSuspender.using(surface, () -> {
            dataSeries.updateRangeXyzAt(0, xValues, yValues, zValues);
        });
    };
    // </UpdateData>

    private Double getGaussianRandomNumber(Double mean, Double stdDev) {
        final Random random = new Random();

        final double u1 = random.nextDouble();
        final double u2 = random.nextDouble();

        final double randStdNormal = Math.sqrt(-2.0 * Math.log(u1)) * Math.sin(2.0 * Math.PI * u2);
        return mean * stdDev * randStdNormal;
    }
}

File: samples/tutorials-native/tutorials-3d/tutorial-4/kotlin/src/main/java/com/scichart/tutorial/MainActivity.kt
package com.scichart.tutorial

import android.os.Bundle
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import com.scichart.charting.visuals.SciChartSurface
import com.scichart.charting3d.model.dataSeries.xyz.XyzDataSeries3D
import com.scichart.charting3d.modifiers.*
import com.scichart.charting3d.visuals.SciChartSurface3D
import com.scichart.charting3d.visuals.axes.NumericAxis3D
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D
import com.scichart.charting3d.visuals.renderableSeries.scatter.ScatterRenderableSeries3D
import com.scichart.core.framework.UpdateSuspender
import com.scichart.core.model.DoubleValues
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import kotlin.math.ln
import kotlin.math.sin
import kotlin.math.sqrt

class MainActivity : AppCompatActivity() {
    // <DataSeriesSetup>
    private val pointsCount = 200

    private val scheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private lateinit var schedule: ScheduledFuture<*>

    private lateinit var surface: SciChartSurface3D

    private val xValues = DoubleValues()
    private val yValues = DoubleValues()
    private val zValues = DoubleValues()

    private val dataSeries = XyzDataSeries3D(Double::class.javaObjectType, Double::class.javaObjectType, Double::class.javaObjectType)
    // </DataSeriesSetup>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        SciChartSurface.setRuntimeLicenseKey("")

        val chartLayout = findViewById<ViewGroup>(R.id.chart_layout)

        surface = SciChartSurface3D(this)
        chartLayout.addView(surface)

        // <AppendDataSeries>
        for (i in 0 until pointsCount) {
            xValues.add(getGaussianRandomNumber(5.0, 1.5))
            yValues.add(getGaussianRandomNumber(5.0, 1.5))
            zValues.add(getGaussianRandomNumber(5.0, 1.5))
        }
        dataSeries.append(xValues, yValues, zValues)
        // </AppendDataSeries>

        val pointMarker = SpherePointMarker3D()
        pointMarker.fill = 0xFF32CD32.toInt()
        pointMarker.size = 10.0f

        val rSeries = ScatterRenderableSeries3D()
        rSeries.dataSeries = dataSeries
        rSeries.pointMarker = pointMarker

        val tooltipModifier = TooltipModifier3D()
        tooltipModifier.crosshairMode = CrosshairMode.Lines

        UpdateSuspender.using(surface) {
            surface.xAxis = NumericAxis3D()
            surface.yAxis = NumericAxis3D()
            surface.zAxis = NumericAxis3D()
            Collections.addAll(surface.renderableSeries, rSeries)
            Collections.addAll(
                surface.chartModifiers,
                OrbitModifier3D().apply {
                    executeOnPointerCount = 2
                },
                ZoomExtentsModifier3D(),
                PinchZoomModifier3D(),
                tooltipModifier
            )
        }

        // <CreateTimer>
        schedule = scheduledExecutorService.scheduleWithFixedDelay(updateData, 0, 10, TimeUnit.MILLISECONDS)
        // </CreateTimer>
    }

    // <UpdateData>
    private val random = Random()

    private val updateData = Runnable {
        for (i in 0 until pointsCount) {
            val xValue = xValues.get(i) + random.nextDouble() - 0.5
            val yValue = yValues.get(i) + random.nextDouble() - 0.5
            val zValue = zValues.get(i) + random.nextDouble() - 0.5

            xValues.set(i, xValue)
            yValues.set(i, yValue)
            zValues.set(i, zValue)
        }

        UpdateSuspender.using(surface) {
            dataSeries.updateRangeXyzAt(0, xValues, yValues, zValues)
        }
    }
    // </UpdateData>

    private fun getGaussianRandomNumber(mean: Double, stdDev: Double) : Double {
        val random = Random()

        val u1: Double = random.nextDouble()
        val u2: Double = random.nextDouble()

        val randStdNormal = sqrt(-2.0 * ln(u1)) * sin(2.0 * Math.PI * u2)
        return mean * stdDev * randStdNormal
    }
}
